autoload/maktaba/json.vim

```diff
@@ -0,0 +1,211 @@
+  if l:remaining[0:0] is# ']'
+    return [[], s:Consume(l:remaining, 1)]
+  endif
+  let l:items = []
+  while !empty(l:remaining)
+    " Parse and consume one value as the next list item.
+    let [l:item, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    call add(l:items, l:item)
+    " If next character is "]", consume and finish list.
+    if l:remaining[0:0] is# ']'
+      let l:remaining = s:Consume(l:remaining, 1)
+      break
+    endif
+    " Only other acceptable character is comma. Consume and continue.
+    if l:remaining[0:0] isnot# ','
+      throw maktaba#error#BadValue(
+          \ 'Junk after JSON array item: %s', s:Ellipsize(l:remaining, 30))
+    endif
+    let l:remaining = s:Consume(l:remaining, 1)
+  endwhile
+  return [l:items, l:remaining]
+endfunction
+
+function! s:ParseDictPartial(json, custom_values) abort
+  let l:remaining = s:Consume(a:json, 1)
+  if l:remaining[0:0] is# '}'
+    return [[], s:Consume(l:remaining, 1)]
+  endif
+  let l:dict = {}
+  while !empty(l:remaining)
+    " Parse and consume one value as the key.
+    let [l:key, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    if !maktaba#value#IsString(l:key)
+      throw maktaba#error#BadValue(
+          \ 'Non-string keys not allowed for JSON dicts: %s.', string(l:key))
+    endif
+    " Consume separating ":".
+    if l:remaining[0:0] isnot# ':'
+      throw maktaba#error#BadValue(
+          \ 'Junk after JSON key: %s', s:Ellipsize(l:remaining, 30))
+    endif
+    let l:remaining = s:Consume(l:remaining, 1)
+    " Parse and consume one value as the item value.
+    let [l:value, l:remaining] = s:ParsePartial(l:remaining, a:custom_values)
+    let l:dict[l:key] = l:value
+    unlet l:value
+
+    " If next character is "}", consume and finish list.
+    if l:remaining[0:0] is# '}'
+      let l:remaining = s:Consume(l:remaining, 1)
+      break
+    endif
+    " Only other acceptable character is comma. Consume and continue.
+    if l:remaining[0:0] isnot# ','
+      throw maktaba#error#BadValue(
+          \ 'Junk after JSON object item: %s', s:Ellipsize(l:remaining, 30))
+    endif
+    let l:remaining = s:Consume(l:remaining, 1)
+  endwhile
+  return [l:dict, l:remaining]
+endfunction
+
+function! s:SetDifference(a, b) abort
+  let l:difference = []
+  for l:a_item in a:a
+    if index(a:b, l:a_item) == -1
+      call add(l:difference, l:a_item)
+    endif
+  endfor
+  return l:difference
+endfunction
+
+""
+" Parses the JSON text {json} to a Vim value. If [custom_values] is passed, it
+" is a dictionary mapping JSON primitives (in string form) to custom values to
+" use instead of maktaba#json# sentinels. For example: >
+"   let value = maktaba#json#Parse('[null]', {'null': ''})
+" <
+" @throws WrongType if {json} is not a string or [custom_values] is not a dict.
+" @throws BadValue if {json} is not a valid JSON text.
+" @throws BadValue if [custom_value] keys are invalid JSON primitive names.
```

Comment:
typo: [custom_values], right? (I assume that's why the function prototype looks odd in the help.)